docs/sources/lids/_DRAFT_QueryFairness.md

```diff
@@ -0,0 +1,319 @@
+This should also include changes to the queue structure to be easily extensible for future scheduling improvements. 
+
+## Non-Goals (optional)
+
+While changing and extending the scheduler requires also user-facing API changes, the public API is not part of the discussion of this document.
+
+## Proposals
+
+### Proposal 0: Do nothing
+
+An alternative to changing the scheduling mechanism is to handle QoS control via multiple tenants and multi-tenant querying.
+
+**Pros:**
+* Keeps the scheduler as simple as it is now
+* No development time
+
+**Cons:**
+* While that separation into tenants may work for some prospects, it might not be feasible to implement for others.
+
+### Proposal 1: Add fixed second level to scheduler
+
+The scheduler is implemented in a way that it maintains a separate FIFO queue for each tenant. When a request (sub-query) is enqueued, the scheduler puts it into the existing queue for that tenant. If the queue does not exist yet, it creates it first and re-assignes the connected querier workers to the available tenant queues. Each querier worker pulls round-robin from the assigned queues in a loop.
+
+Now, instead of enqueuing and pulling directly from the per-tenant queue, requests get enqueued in per-user queues and the per-tenant queue pulls round-robin from the user queues that are assigned to the tenant queues.
+
+**Component diagram:**
+
+```plantuml
+skinparam componentStyle rectangle
+
+package Frontend {
+    () RoundTripper as QF
+}
+
+package Scheduler {
+    () FrontendLoop as FL
+
+    component RequestQueue {
+        [User Queue] as U1
+        [User Queue] as U2
+        [User Queue] as U3
+        [User Queue] as U4
+        [Tenant Queue] as T1
+        [Tenant Queue] as T2
+    }
+
+    () QuerierLoop as Q1
+    () QuerierLoop as Q2
+
+    U1 ..>> T1
+    U2 ..>> T2
+    U3 ..>> T2
+    U4 ..>> T2
+}
+
+QF -->> FL
+FL -->> U1
+FL -->> U2
+FL -->> U3
+FL -->> U4
+
+package Querier {
+    [Worker] as W1
+    [Worker] as W2
+}
+
+T1 -->> Q1
+Q1 -->> W1
+
+T2 -->> Q1
+T2 -->> Q2
+Q2 -->> W2
+```
+
+![component diagram](../scheduler-proposal-1-component-diagram.png)
+
+Like the current implementation, the scheduler enqueues requests based on the `X-Scope-OrgID` header (or equivalent key in the request context), but also takes a second key (such as `X-Scope-UserID`) into account. This ensembles a fixed hierarchy with two levels where the tenant-to-user relation is a one-to-many relation.
+However, this has the disadvantage that the concept of users (that does not exist yet in Loki) leaks into the scheduler domain.
+
+**Pros:**
+* Relatively simple to to implement
+
+**Cons:**
+* Not extensible
+* Leaks domain knowledge
+
+### Proposal 2: Fully hierarchical scheduler
```

Comment:
This is easier to implement IMHO. It abstracts away the nesting neatly and we would be able to control it with a feature flag.